bitswap/server/internal/decision/engine.go

```diff
@@ -717,39 +717,72 @@ func (e *Engine) MessageReceived(ctx context.Context, p peer.ID, m bsmsg.BitSwap
 		e.peerLedger.ClearPeerWantlist(p)
 	}
 
-	s := uint(e.peerLedger.WantlistSizeForPeer(p))
-	if wouldBe := s + uint(len(wants)); wouldBe > e.maxQueuedWantlistEntriesPerPeer {
-		log.Debugw("wantlist overflow", "local", e.self, "remote", p, "would be", wouldBe)
-		// truncate wantlist to avoid overflow
-		available, o := bits.Sub(e.maxQueuedWantlistEntriesPerPeer, s, 0)
-		if o != 0 {
-			available = 0
+	if len(wants) != 0 {
+		filteredWants := wants[:0] // shift inplace
+		for _, entry := range wants {
+			if entry.Cid.Prefix().MhType == mh.IDENTITY {
+				// This is a truely broken client, let's kill the connection.
+				e.lock.Unlock()
+				log.Warnw("peer wants an identity CID", "local", e.self, "remote", p)
+				return true
+			}
+			if e.maxCidSize != 0 && uint(entry.Cid.ByteLen()) > e.maxCidSize {
+				// Ignore requests about CIDs that big.
+				continue
+			}
+			filteredWants = append(filteredWants, entry)
+			if len(filteredWants) == int(e.maxQueuedWantlistEntriesPerPeer) {
+				// filteredWants at limit, ignore remaining wants from request.
+				log.Debugw("requested wants exceeds max wantlist size", "local", e.self, "remote", p, "ignoring", len(wants)-len(filteredWants))
+				break
+			}
 		}
-		wants = wants[:available]
-	}
-
-	filteredWants := wants[:0] // shift inplace
-
-	for _, entry := range wants {
-		if entry.Cid.Prefix().MhType == mh.IDENTITY {
-			// This is a truely broken client, let's kill the connection.
-			e.lock.Unlock()
-			log.Warnw("peer wants an identity CID", "local", e.self, "remote", p)
-			return true
+		wants = wants[len(filteredWants):]
+		for i := range wants {
+			wants[i] = bsmsg.Entry{} // early GC
 		}
-		if e.maxCidSize != 0 && uint(entry.Cid.ByteLen()) > e.maxCidSize {
-			// Ignore requests about CIDs that big.
-			continue
+		wants = filteredWants
```

Comment:
Not sure which line to leave this comment on, but adding some thoughts based on conversations with @gammazero and @Stebalien

It might be helpful if we logically had two "wantlists" here:
1. The set of blocks that the server is currently in the middle of processing responses for (i.e. request-response style semantics)
2. The set of blocks that the server doesn't currently have, but if it receives them it will notify / send them out to requesters (i.e. subscription semantics)

We currently have the wantlist and taskqueue which might be these two queues, but they also might not given we want to be able to do things like cancel tasks in the taskqueue without needing to re-enumerate the entire queue in case it's large. It might also be fine if the capacity for subscription-wants only really exists if the request-response wants have been satisfied already (e.g. if there are tons of new requests coming in flushing out all the old subscriptions is probably fine, but it seems ok to have more capacity for request-response which is supposed to be "moving" than for subscriptions that can stay and take up memory indefinitely).

Also, for others watching: I think it's fairly clear that some protocol changes to help with backpressure here would be great, but since they're protocol changes those will have to wait for another time 😅.

Reply:
We use Bitswap with a case where more than 1024(default max) CIDs are requested in a single `GetBlocks`. If we do that, the Bitswap hangs because of this limit and recovers in a minute after rebroadcasting. It would be great if Bitswap clients could handle that case and avoid sending more than protocol-wide constant by backpressure-ing the caller.

Reply:
> It would be great if Bitswap clients could handle that case and avoid sending more than protocol-wide constant by backpressure-ing the caller.

Could you clarify your suggestion? Reading it I see 3 options:
1. Having the server backpressure the client on the other side of the wire
   - Similar to my comment above, and AFAICT requires some protocol changes to accommodate
2. Have the bitswap client internally batch `GetBlocks` calls that are for larger than `maxWantlistSize` into batches before returning them in case the batch is fully returned before the rebroadcast interval?
   - Doable if it'll be helpful, although it's a bit gross since A) `maxWantlistSize` shouldn't really be a protocol/network-wide thing and be per-client B) this batching can be done outside of the bitswap client package.
3. The bitswap client backpressuring the caller (e.g. code that's walking a DAG)
   - My suspicion is this would best be served by having a streaming version of `GetBlocks` that you could block on which is a separate problem that seems like https://github.com/ipfs/boxo/issues/121. Which if you're interested in let's chat on the issue there.

Reply:
Thank you for the elaborate response! 

My suggestion was some form of 2 and 3. The client should be able to deal with the server-side request rate-limiting; otherwise, the client gets stuck expecting to be served, while the server simply cuts off his wants. The rebroadcasting after one minute helps, but that's still one minute and request spamming overhead over the wire. 

Ideally, we would do a protocol change as described in 1, but as it's breaking, we may consider other less clean options, like or similar to 2. Setting protocol-wide `maxWantlistSize` is gross, I agree. Another option might be negotiating the limit between the client and the server so the client knows it should never exceed it.

The 3 is complimentary and provides a new, powerful way to interface with a client. However, I just realized that it is not necessary for our case if the client is smart enough. In our case, we have a flat structure, i.e. we don't traverse a DAG where we unpack IPLD nodes to get more CID links to fetch them and unpack again to get to the data. Essentially, we know all the CIDs in advance and could simply ask Bitswap to get all of them over `GetBlocks` as long the client is smart enough not to get into any limitations of its immediate peers, which we are currently facing with this issue.

Reply:
Re protocol breaking 1. We are actually fine with this being a protocol-breaking change as we are building a new Bitswap-based protocol that hasn't been deployed yet and I believe there is more or less a clear way how to handle protocol version bumps in bitswap network component.

Reply:
@Wondertan With the latest changes in this PR (as of the date of this comment), the main problem should be solved. The problem was that wants that have DONT_HAVE responses remain on the queue, and cause new incoming messages to be dropped because there is no more free space on the queue, even if those new messages could update existing wants. Incoming wants are now applied to the queue, and any new ones that cannot be handled without increasing the queue over the size limit are handled specially as overflow, where their priority is considered for possibly replacing existing wants. Please take a look at the latest.

I still think that having the client limit its message sizes will be useful to avoid oversized wantlists from being truncated and allowing any overflow is handled. The client can also group batches of wants by priority if applicable. If the future, priority can be influenced by dag path (higher priority for items closer to dag root).

Reply:
This conversation moved to issue #644